src/nvim/os/keyboard.c

```diff
@@ -0,0 +1,250 @@
+///
+/// To be used in os_mods_status() function.
+///
+/// @param[out]  mods  Holds the information about the status of various locks
+///                    in form of a bitmask
+/// @param[out]  err  Location where error message is to be saved, set to NULL
+///                   if no error.
+///
+/// @return true in case of error, false otherwise.
+static bool os_get_locks_status(int *const mods, const char **const err)
+{
+  *err = NULL;
+  io_service_t ios;
+  io_connect_t ioc;
+  CFMutableDictionaryRef mdict;
+  bool stateCaps, stateNums;
+  mdict = IOServiceMatching(kIOHIDSystemClass);
+  ios = IOServiceGetMatchingService(kIOMasterPortDefault,
+                                    (CFDictionaryRef)mdict);
+  if (!ios) {
+    if (mdict) {
+      CFRelease(mdict);
+    }
+    *err = "Unable to start service to query data";
+    return false;
+  }
+
+  if (IOServiceOpen(ios, mach_task_self(), kIOHIDParamConnectType,
+                    &ioc) != kIOReturnSuccess) {
+    IOObjectRelease(ios);
+    *err = "Unable to query data";
+    return false;
+  }
+  IOObjectRelease(ios);
+
+  if (IOHIDGetModifierLockState(ioc, kIOHIDCapsLockState,
+                                &stateCaps) != kIOReturnSuccess) {
+    *err = "Unable to lock the state";
+    return false;
+  }
+
+  if (IOHIDGetModifierLockState(ioc, kIOHIDCapsLockState,
+                                &stateNums) != kIOReturnSuccess) {
+    *err = "Unable to lock the state";
+    return false;
+  }
+  if (stateCaps) {
+    *mods = *mods | kCapsLock;
+  }
+  if (stateNums) {
+    *mods = *mods | kNumLock;
+  }
+  IOServiceClose(ioc);
+  return true;
+}
+#else
+/// A dummy function to be used by os_mod_status() if none of the above
+/// platforms were detected
+///
+/// @param[out]  mods  Holds the information about the status of various locks
+///                    in form of a bitmask. Sets to 0 for dummy
+/// @param[out]  err  Location where error message is to be saved, set to NULL
+///                   if no error. No error detected for dummy
+///
+/// @return true for all the cases
+static bool os_get_locks_status(int *const mods, const char **const err)
+{
+  *mods = 0;
+  *err = NULL;
+  return true;
+}
+#endif
+
+/// Get information about Num/Caps/Scroll Lock state
+///
+/// To be used in nvim_get_keyboard_mods_state() function.
+///
+/// @param[out]  dict  Pointer to dictionary where information about modifiers
+///                    is to be dumped.
+/// @param[out]  err  Location where error message is to be saved, set to NULL
+///                   if no error.
+///
+/// @return true in case of error, false otherwise.
+bool os_mods_status(Dictionary *const dict, const char **const err)
+{
+  int mods = 0;
+  *err = NULL;
+  if (!os_get_locks_status(&mods, err)) {
+      return false;
+  }
+  PUT(*dict, "CapsLock", BOOLEAN_OBJ(mods & kCapsLock));
+  PUT(*dict, "NumLock", BOOLEAN_OBJ(mods & kNumLock));
+  PUT(*dict, "ScrollLock", BOOLEAN_OBJ(mods & kScrollLock));
```

Comment:
This will yield three false values for the `#else` case. It should yield empty dictionary instead, *for querying lock states being not supported*. And I do not see scroll lock support in OS X code as well, so this key also must be out in this case. This is exactly what I was talking about when I said about returning mask.

Reply:
OK I make the changes

Reply:
This makes the sense to me now